src/Workspaces/Remote/ServiceHub/Shared/ClientDirectStream.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.IO.Pipes;
+using System.Runtime.Remoting;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    /// <summary>
+    /// Direct stream between server and client to pass around big chunk of data
+    /// </summary>
+    internal class ClientDirectStream : Stream
+    {
+        // 128KB buffer size
+        private const int BUFFERSIZE = 4 * 1024;
+
+        private readonly string _name;
+        private readonly NamedPipeClientStream _pipe;
+        private readonly Stream _stream;
+
+        public ClientDirectStream(string name)
+        {
+            // this type exists so that consumer doesn't need to care about all these arguments/flags to get good performance
```

Comment:
well, then you need to have both pipe and buffer stream since we need wait for connection and wait for server stuff.

so, without this type, all user has to do

using (pipe = new NamedPipe ... )
using (stream = new BufferedStream(pipe))
{
        pipe.WaitForXXXX
        writer.WriteTo( ... stream)
}

this basically, package two to one.